[HTTP] Remove unnecessary clones by passing around owned data

GnomedDev opened the pull request:
A lot of the methods on Request and AttachmentType were only given references or mutable references to their data, instead of taking owned data. This was unnecessary and caused a lot of cloning. This PR fixes this

nickelc committed:
Remove deprecated `audit_log` re-exports and types (#1864)

The audit log types are no longer re-exported in the `model::guild` module.

BREAKING CHANGES:
- The `audit_log::Action*` enum sub types are renamed to `*Action`.
  For instance, `ActionChannel` to `ChannelAction`.

- The audit log types must be imported directly from the
  `model::guild::audit_log` module.

GnomedDev committed:
Reference the objects within `ResolvedTarget` instead of cloning them (#1885)

nickelc committed:
Replace `gateway::WsStream` & extension traits with a newtype struct (#1893)

The newtype struct `gateway::WsClient` hides the WebSocket
implementation details and is the first step to remove the
`async_tungstenite` & `tungstenite` types from the public API.

The `WebSocketGatewayClientExt`, `ReceiverExt` and `SenderExt` trait
methods are merged into the implementation of the newtype struct.

BREAKING CHANGES:
- `gateway::WsStream` is replaced by `gateway::WsClient`.
- The `gateway::WebSocketGatewayClientExt` trait is removed.

Miezhiko committed:
Change `Typing::stop` to return a `bool` (#1897)

This changes the return type of `http::Typing::stop` to be a simple
`bool` rather than `Option<()>`, which doesn't provide useful
information that couldn't be expressed with just a `bool`.

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>
Signed-off-by: Miezhiko <Miezhiko@gmail.com>

nickelc committed:
Remove useless reference on `Option<&T>` (#1896)

Since `&` is Copy, it’s useless to have a reference on `Option<&'a [u8]>`.

nickelc committed:
Remove deprecated interaction type aliases, field and method (#1902)

GnomedDev committed:
Switch to `Into<String>` when accepting string parameters (#1903)

This changes instances of `ToString` to `Into<String>` whenever ownership of
the `String` is necessary. In case it is not, it's simply changed to `&str` to
lessen the effects of monomorphization on compile times.

vaporoxx committed:
Allow having no framework when the `framework` feature is enabled (#1908)

vaporoxx committed:
Do not set message content not sent by the API (#1910)

Problems with the current setup:

- The content of messages with these types is not documented, meaning that it can change anytime.
- Not every message type is handled, e.g. boost or thread creation messages do not get their content added.
- Overall more possibly unexpected behaviour and less consistency with the API.

GnomedDev committed:
Remove unneeded `futures::iter::stream` usage (#1922)

mkrasnitski committed:
Add support for executing webhooks in threads (#1923)

This adds a method `Webhook::execute_in_thread` which takes an additional
thread ID (which is just a `ChannelId`) parameter. Then, the webhook will
execute in that thread, provided it belongs to the webhook's assigned guild
channel.

This also adds a `thread_id` parameter to `Http::execute_webhook`, which is a
breaking change.

GnomedDev committed:
Add a cache fallback to `ChannelId::message` (#1924)

GnomedDev committed:
Change `User::has_role` to accept just a `GuildId` (#1927)

This also removes `GuildContainer`, as this was (and historically has been) the only use of it.

GnomedDev committed:
Implement suggested `TODO` fixes (#1928)

GnomedDev committed:
Expose `dashmap`'s wrapped `Ref` to reduce cloning (#1917)

GnomedDev committed:
Move around `Box`es to fix `clippy::large_enum_variant` (#1929)

GnomedDev committed:
Fix clippy warnings (#1933)

GnomedDev committed:
Avoid cloning member out of cache in `Guild::member` (#1934)

GnomedDev committed:
De-duplicate cache -> http logic from `Message::member` (#1935)

mkrasnitski committed:
Clean up previous changes to cache (#1936)

This replaces the implementation of `From<Ref>` for `CacheRef` with a
private method in order to not expose the dashmap type as part of the
public API.

Also, this rewords and removes a few clones on types returned from the cache.

vaporoxx committed:
Remove references from `ShardManagerOptions` (#1937)

nickelc committed:
Unasync `ShardManager::new` (#1940)

nickelc committed:
Introduce a separate data type for autocomplete interactions (#1941)

The interaction payload for autocompletes is slightly different. It's
missing the `resolved` data map and reports only for the current
autocompleted option value the `focused` field and its value as a string.

The focused option value is modeled as an enum variant of `CommandDataOptionValue`.

```
enum CommandDataOptionValue {
    Autocomplete { name: String, kind: CommandOptionType, value: String },
}
```

The focused option value can be retrieved via `AutocompleteData::focused_option()`
from the `AutocompleteData::options` vector.

GnomedDev committed:
Reduce int size of `prune_count` methods (#1947)

GnomedDev committed:
Clean and optimise `message_builder.rs` (#1946)

GnomedDev committed:
Remove unused allowed clippy warnings or fix them (#1945)

GnomedDev committed:
Use concrete fields instead of `HashMap` in the `GetMessages` builder (#1942)

GnomedDev committed:
Remove user-only `EditProfile` methods (#1959)

BREAKING CHANGE: This removes methods only usable on user accounts.

GnomedDev committed:
Improve event deserialization by sacrificing `GuildUnavailable` & `Unknown` events (#1938, #1951)

BREAKING CHANGES: Removes the `guild_unavailable` & `unknown` methods
from the `EventHandler` trait.

GnomedDev committed:
Rework manual deserialization implementations (#1938, #1951)

GnomedDev committed:
Fix `needless_pass_by_value` caused by new `ToString` usage (#1961)

GnomedDev committed:
Store `Vec<u8>` instead of `&[u8]` in `Request` (#1963)

GnomedDev committed:
Remove `kind` from the interaction types (#1970)

nickelc committed:
Remove unneeded `#[repr(u8)]` attributes (#1972)

GnomedDev committed:
Accept `Into<String>` in more places (#1955)

nickelc committed:
Add missing `member` field to `MessageInteraction` (#1976)

nickelc committed:
Rename `Permissions::USE_SLASH_COMMANDS` to `USE_APPLICATION_COMMANDS` (#1977)

The permission was renamed by Discord in regards to the new context menu commands.

GnomedDev committed:
Rework builder module to get rid of all `HashMap<'static str, T>`  (#1962)

vaporoxx committed:
Update `OpCode` to match Discord's documentation (#1980)

GnomedDev committed:
Add missing `#[serde(rename = "type")]` attributes (#1987)

GnomedDev committed:
Remove usage of the `json!` macro from websocket message sending (#1981)

GnomedDev committed:
Use `ReactionType` to serialize the emoji in `CreateComponents` (#1989)

GnomedDev committed:
Reuse `edit_member` in `GuildId` methods (#1991)

GnomedDev committed:
Use custom enum for `Message::nonce` instead of `json::Value` (#1990)

GnomedDev committed:
Add proper builders for `Webhook` (#1993)

This adds `CreateWebhook` and `EditWebhook`, and gets rid of the specific
methods for different combinations of arguments. This also gets rid of the
`json!` usage and `Value` serialization, instead using `derive(Serialize)`

GnomedDev committed:
Swap all ID types to use `NonZeroU64` internally (#1930)

GnomedDev committed:
Use `CacheRef` for `CurrentUser` (#1996)

This removes `Cache::current_user_id` and `current_user_field` and makes
`Cache::current_user` return a `CacheRef` by reworking the internals of
`CacheRef` to hold an enum of different guard types, currently only a dashmap
`Ref` or a `parking_lot::RwLockReadGuard`. This prevents unnecessary cloning of
the current user and monomorphization bloat from the `_field` method.

GnomedDev committed:
Store messages in a `HashMap` instead of a `DashMap` (#1995)

The message cache is currently stored as `DashMap<ChannelId, DashMap<MessageId,
Message>>` however since the outer `DashMap` already provides interior
mutability, it doesn't need another `DashMap` inside. 

This also stops dashmap being exposed to the public API via
`MessageIterator::Item` by removing it (and instead exposes the inner
`HashMap`).

GnomedDev committed:
Deduplicate a lot of the collector module (#1932)

nickelc committed:
Group & sort derive macros alphabetically

The derive macros are grouped into `std` & third party and
sorted alphabetically.

```
struct Foo;
```

nickelc committed:
Unify `derive` & `#[non_exhaustive]` attributes for number enums (#1999)

BREAKING CHANGE: Some enums were missing the `[non_exhaustive]` attribute.

nickelc committed:
Remove unused `channel/sticker.rs` file

The file was forgotten to delete when the types were moved into the
root. See 34bc5fec302c1f66cc851d6626f04a8831c4f8b2

nickelc committed:
Replace magic number with `ChannelType::Stage`

nickelc committed:
Group & sort derive macros alphabetically

The derive macros are grouped into `std` & third party and
sorted alphabetically.

```
struct Foo;
```

GnomedDev committed:
Use `CacheRef` to access channel messages from the cache (#2000)

BREAKING CHANGE: `Cache::channel_messages_field` is replaced by `Cache::channel_messages`
returning a reference to the messages map.

nickelc committed:
Make `MessageActivityKind`'s variants upper camel case (#2005)

GnomedDev committed:
Reduce size of shard info and turn it into a struct (#1984)

This replaces the `[u64; 2]` shard info with `ShardInfo { id: u32, total: u32 }`.

Co-authored-by: nickelc <constantin.nickel@gmail.com>

GnomedDev committed:
Add `UserRef` by wrapping temporary cache values in `Arc` (#1997)

vaporoxx committed:
Rework setting and storing of presence data (#2007)

nickelc committed:
Preserve values for unknown enum variants (#2008)

The `enum_number!` macro now takes the whole enum definition and generates
`From` trait implementations to convert a value to the enum and back.

The implementations can then be picked up by `serde` with
`#[serde(from = "u8", into = "u8")]` to (de)serialize the types.

```
enum_number! {
    #[derive(Clone, Copy, Debug, Deserialize, Serialize)]
    #[serde(from = "u8", into = "u8")]
    pub enum Foo {
        A = 1,
        B = 2,
        _ => Unknown(u8),
    }
}
```

BREAKING CHANGE: The `Unknown` variant now takes the unknown value
and the removed `fn num() -> u64` method can be replaced
with `let v = u8::from(kind)`.

nickelc committed:
fixup! Use `CacheRef` for `CurrentUser` (#1996)

GnomedDev committed:
Remove `Shard` prefix in `ShardInfo` logging (#2014)

Before `ShardInfo`:
`[serenity::gateway::shard]: [Shard [15, 33]] ...`

Currently:
`[serenity::gateway::shard]: [Shard ShardInfo { id: 15, total: 33 }] ...`

After:
`[serenity::gateway::shard]: [ShardInfo { id: 15, total: 33 }] ...`

nickelc committed:
Remove unneeded `#[must_use]` attribute

The returned `Result` is already marked as `#[must_use]`.

GnomedDev committed:
Clean up usage of `ref`, `ref mut`, and `as_ref()` (#2016)

GnomedDev committed:
Change `Member::edit` to edit the member in-place (#2017)

GnomedDev committed:
Fix possible deadlock in `PartialGuild::greater_member_hierarchy` (#2015)

GnomedDev committed:
Remove `InterMessage::Json` (#2009)

vaporoxx committed:
Support setting presence when identifying (#2018)

GnomedDev committed:
Clean up collector module by replacing duplication with generics (#2011)

nickelc committed:
Use `GenericId` for the `AuditLogEntry::target_id` field (#2023)

GnomedDev committed:
Fix return type for the guild delete endpoint (#2030)

nickelc committed:
Replace `json::from_number` with `into()` (#2034)

Both `Value` types implement `From` for their number variants.

nickelc committed:
Merge application command & autocomplete interactions (#2031)

The structs are almost identical except for the `focused` field for the
current autocompleted option and that the `resolved` data map is not
sent for autocomplete interactions.

The entities from the `CommandData::resolved` data map are no longer
cloned for the option values during deserialization. But the resolved
options and values can be accessed via the `CommandData::options()` method.

For autocomplete interactions `CommandData::autocomplete()` returns the
current autocompleted option + value.

BREAKING CHANGES:
- `ApplicationCommandInteraction` is now used for application commands
  and their autocomplete interaction.

- The `CommandDataOptionValue`'s variants no longer have the cloned
  entities from the `CommandData::resolved` data map.
  See the `CommandData::options` method.

- The fields `value`, `kind`, `options` and `focused` of `CommandDataOption`
  are streamlined into the `CommandDataOptionValue` enum with new variants
  for subcommands, subcommand groups and autocomplete.

nickelc committed:
Switch to `tokio-tungstenite` (#2036)

GnomedDev committed:
Use `try_from` and `into` serde attributes to deserialize `PermissionOverwrite` (#2042)

GnomedDev committed:
Fix `Timestamp` having a fallible `From` impl (#2032)

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

GnomedDev committed:
Constify some functions that could be used in a `const` context (#2047)

GnomedDev committed:
Fix new clippy warnings (#2041)

arqunis committed:
Fix deserialisation for reaction events

To simplify the deserialisation implementation for `Event`, it was
changed to use serde's native support for `enum` representations that
are adjacently tagged (meaning in a text form like JSON, the tag is in
one field, the data in the other field).

This helped in trimming the lines of code we have to maintain; however,
it did break reaction events whose names are prepended `MESSAGE_`, which
serde could not account for when renaming, for instance, `ReactionAdd`
to SCREAMING_SNAKE_CASE -> `REACTION_ADD`, when it should have been
`MESSAGE_REACTION_ADD`.

This commit manually renames the reaction events to have `MESSAGE_` at
the beginning of their event names.

GnomedDev committed:
Clean up dispatch and unnecessary clones in client (#2019)

GnomedDev committed:
Make standard framework configurable at runtime (#1950)

GnomedDev committed:
Deduplicate `Filter` in collector (#2040)

arqunis committed:
Revert "Clean up dispatch and unnecessary clones in client (#2019)"

This reverts commit de05e5bb3a99b45010b5aa2ebaa64a32db2d2e97 because it
broke shard code to the point where it just crashes, preventing from
even a simple ping-pong bot from running.

Miezhiko committed:
Increase integer range of `GuildChannel::message_count` from `u8` to `u32` (#2054)

GnomedDev committed:
Deserialize to `JsonMap` then recreate `Value` when needed (#2057)

GnomedDev committed:
Use `NonZeroU32` in `CommonFilterOptions` (#2052)

arqunis committed:
Fix compilation errors that arose from rebasing

Additionally, adapt some code to the `next` branch.

GnomedDev committed:
Fix clippy warnings and apply `next` patches to new code (#2061)

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

mkrasnitski committed:
Overhaul usage of the builder pattern (#2024)

GnomedDev committed:
Remove unnecessary generics (#2065)

GnomedDev commented:
Looking at the tests that failed, they are related to the AttachmentType `From` impl changing, but taking a second look it seems to be quite ambiguous anyway, should I remove these conversions and make it explicit for the user?

mkrasnitski commented on src/model/channel/attachment_type.rs:
```suggestion
#[cfg(feature = "http")]
async fn data_path(path: &Path) -> Result<Vec<u8>> {
```

You should feature-gate these, as they're only used if `http` is enabled (would also fix the import error).

mkrasnitski commented on src/model/channel/attachment_type.rs:
```suggestion
#[cfg(feature = "http")]
async fn data_path(path: &Path) -> Result<Vec<u8>> {
```

You should feature-gate this, as it's only used if `http` is enabled (would also fix the import error).

mkrasnitski commented on src/model/channel/attachment_type.rs:
```suggestion
#[cfg(feature = "http")]
async fn data_file(file: &File) -> Result<Vec<u8>> {
```

mkrasnitski commented on src/model/channel/attachment_type.rs:
```suggestion
#[cfg(feature = "http")]
async fn data_image(client: &Client, url: Url) -> Result<Vec<u8>> {
```

mkrasnitski commented on src/http/multipart.rs:
As far as I understand it, nowhere in the lib is a multipart actually re-used, since everything takes `Request`s by ownership instead of borrowing?

mkrasnitski commented on src/model/channel/attachment_type.rs:
Github won't let me suggest changes directly here, but you can change this to take an `Into<Cow>` as follows:

```rust
impl<'a, S> From<(&'a File, S)> for AttachmentType<'a>
where
    S: Into<Cow<'static, str>>,
{
    fn from((file, filename): (&'a File, S)) -> AttachmentType<'a> {
        AttachmentType::File {
            file,
            filename: filename.into(),
        }
    }
}
```

mkrasnitski commented on src/model/channel/attachment_type.rs:
Same as above, I'd recommend replacing as follows:
```rust
impl<'a, S> From<(Cow<'static, [u8]>, S)> for AttachmentType<'a>
where
    S: Into<Cow<'static, str>>,
{
    fn from((data, filename): (Cow<'static, [u8]>, S)) -> Self {
        AttachmentType::Bytes {
            data,
            filename: filename.into(),
        }
    }
}
```
Btw, why did you change it to `AttachmentType<'static>`, was it to coerce `AttachmentType::Bytes(&'static [u8], &'static str)` to be `'static` as well? I'd say with the new impl it's better to loosen the lifetime again.

mkrasnitski commented on src/model/channel/attachment_type.rs:
Note that it's currently not possible to change the `Cow<'static, [u8]>` into an `impl Into<Cow>` because the compiler complains about conflicting implementations, citing:
```
upstream crates may add a new impl of trait `From<&tokio::fs::File>` for type `Cow<'static, [u8]>` in future versions
```
Now, I think it's obvious this won't happen, but I don't really like the ergonomics as is, where the user is forced to call `.into()` on their arguments (which normally would be passed inline). I'm in favor of taking a simple `&[u8]` here and collecting into a Vec, but that would mean that `Cow::<'static, u8>::Borrowed` never actually ends up being constructed, so better then to just change the `data` field of the `Bytes` variant to a `Vec<u8>` to avoid confusion.

GnomedDev commented on src/http/multipart.rs:
Yes

mkrasnitski committed:
Fix some builders returning unsendable Futures (#2069)

Due to how the permissions lookups were being performed, the Futures
were being marked `!Send` because `GuildRef` is also `!Send`. So, we
unfortunately have to clone the guild we get from the cache.

GnomedDev committed:
Move `Colour` to `models` and remove feature guards (#2073)

Bloectasy committed:
Fix compilation of `Member::kick_with_reason` caused by `CacheRef` (#2072)

GnomedDev committed:
Reduce unknown event log to debug level (#2064)

mkrasnitski committed:
Remove unneeded and deprecated application commands builders. (#2078)

GnomedDev committed:
Avoid using `serde_json` in the HTTP client when `simd_json` is enabled (#2083)

mkrasnitski committed:
Remove `CreateSelectMenuOptions` builder (#2086)

GnomedDev committed:
Swap to `FxHash` instead of default hasher (#2084)

GnomedDev committed:
Remove `async` from methods that can just be synchronous (#2090)

mkrasnitski committed:
Enforce required fields for builders (#2087)

This commit adds enforcement for unconditionally required fields on builders.
It does so by removing the implementation of `Default` on those builders and
adding a `new` method that takes the required fields as argument.

mkrasnitski committed:
Fix some documentation comments and typos (#2095)

arqunis committed:
Fix issues from rebase

mkrasnitski committed:
Use fallback endpoints in `Webhook::{edit,delete}` if no token is set (#2104)

If the `Webhook::token` field is `None` when these functions are called,
attempt to use the alternate authenticated endpoint instead of erroring.
Also, add audit_log_reason parameter to `Http::edit_webhook_with_token`.

mkrasnitski committed:
Remove needless borrows of `impl AsRef<Http>` and `impl CacheHttp` (#2108)

mkrasnitski committed:
Change some model methods to take `&self` instead of `self` (#2106)

mkrasnitski committed:
Add missing audit log reason support to `Http` and relevant builders (#2105)

GnomedDev committed:
Remove `Arc` around `CacheAndHttp` (#2094)

kangalio committed:
Make `AutocompleteChoice` constructable (#2109)

oSumAtrIX committed:
Fix incorrect documentation for `GuildChannel::create_permission`

mkrasnitski committed:
Elide lifetimes on model methods (#2111)

mkrasnitski committed:
Make `description` field optional when creating slash commands (#2110)

kangalio committed:
Remove `Channel::Category` (#2128)

kangalio committed:
Alter `ChannelId::name` to fetch from HTTP when necessary (#2139)

kangalio committed:
Change channel/role position and bitrate type to `u32` (#2149)

kangalio committed:
Stop changing `EventHandler` signatures depending on the `cache` feature (#2144)

kangalio committed:
Stop considering guild emojis as mentions (#2161)

kangalio committed:
Add `EventHandler::shards_ready` (#2164)

kangalio committed:
Remove `parse_token` (#2162)

It was broken due to tokens not being an officially documented format, and the
format having changed silently since being implemented. `validate_token` still
exists, but it does less intrusive check.

kangalio commented:
Clean diff here: https://github.com/serenity-rs/serenity/compare/5edbf0b3a720c528394a5d88325cb2ec804ae31b...GnomedDev:serenity:cleanup-http

Currently blocked on @GnomedDev reviewing https://github.com/GnomedDev/serenity/pull/1 I believe?

kangalio committed:
Make `CreateButton` and `Button` more type-safe (#2166)

GnomedDev committed:
Remove unnecessary clones by passing around owned data

GnomedDev committed:
Make RouteInfo Copy

GnomedDev committed:
Fmt

GnomedDev commented:
Sorry for the close and reopen, was accidental.

I've just removed most of the public AttachmentType changes, apart from changing the `Cow<'a, [u8]>` to `Vec<u8>` as previously it was useless (but the From implementation still uses Cow, someone else can bikeshed them).

GnomedDev committed:
Feature guard Cow use

mkrasnitski reviewed (changes requested):
I think it's fine to leave it as a `Cow<'a, [u8]>` as it'll cause fewer merge conflicts with #2163. There's no harm in leaving it. Other than that, this looks fine.

GnomedDev commented:
There is harm, it is misleading currently as if it is passed through it goes into `impl Into<Cow<'static, [u8]>>` which means that the Cow is useless as-is.

mkrasnitski commented:
Not sure what you mean, could you give an example?

mkrasnitski commented:
@GnomedDev now that #2163 has been merged, could you rebase this on top and remove the attachments changes?
